pkg/systemdparser/unitfile.go

```diff
@@ -0,0 +1,859 @@
+package systemdparser
+
+import (
+	"fmt"
+	"io"
+	"math"
+	"os"
+	"os/user"
+	"path"
+	"strconv"
+	"strings"
+	"unicode"
+)
+
+/* Code to parse, modify and re-emit system unit files */
+
+type unitLine struct {
+	key       string
+	value     string
+	isComment bool
+}
+
+type unitGroup struct {
+	name     string
+	comments []*unitLine // Comments before the groupname
+	lines    []*unitLine
+}
+
+type UnitFile struct {
+	groups      []*unitGroup
+	groupByName map[string]*unitGroup
+
+	Filename string
+	Path     string
+}
+
+type UnitFileParser struct {
+	file *UnitFile
+
+	currentGroup    *unitGroup
+	pendingComments []*unitLine
+	lineNr          int
+}
+
+func newUnitLine(key string, value string, isComment bool) *unitLine {
+	l := &unitLine{
+		key:       key,
+		value:     value,
+		isComment: isComment,
+	}
+	return l
+}
+
+func (l *unitLine) set(value string) {
+	l.value = value
+}
+
+func (l *unitLine) dup() *unitLine {
+	return newUnitLine(l.key, l.value, l.isComment)
+}
+
+func (l *unitLine) isKey(key string) bool {
+	return !l.isComment &&
+		l.key == key
+}
+
+func (l *unitLine) isEmpty() bool {
+	return len(l.value) == 0
+}
+
+func newUnitGroup(name string) *unitGroup {
+	g := &unitGroup{
+		name:     name,
+		comments: make([]*unitLine, 0),
+		lines:    make([]*unitLine, 0),
+	}
+	return g
+}
+
+func (g *unitGroup) addLine(line *unitLine) {
+	g.lines = append(g.lines, line)
+}
+
+func (g *unitGroup) addComment(line *unitLine) {
+	g.comments = append(g.comments, line)
+}
+
+func (g *unitGroup) prependComment(line *unitLine) {
+	n := []*unitLine{line}
+	g.comments = append(n, g.comments...)
+}
+
+func (g *unitGroup) add(key string, value string) {
+	g.addLine(newUnitLine(key, value, false))
+}
+
+func (g *unitGroup) findLast(key string) *unitLine {
+	for i := len(g.lines) - 1; i >= 0; i-- {
+		l := g.lines[i]
+		if l.isKey(key) {
+			return l
+		}
+	}
+
+	return nil
+}
+
+func (g *unitGroup) set(key string, value string) {
+	line := g.findLast(key)
+	if line != nil {
+		line.set(value)
+	} else {
+		g.add(key, value)
+	}
+}
+
+func (g *unitGroup) unset(key string) {
+	newlines := make([]*unitLine, 0, len(g.lines))
+
+	for _, line := range g.lines {
+		if !line.isKey(key) {
+			newlines = append(newlines, line)
+		}
+	}
+	g.lines = newlines
+}
+
+func (g *unitGroup) merge(source *unitGroup) {
+	for _, l := range source.comments {
+		g.comments = append(g.comments, l.dup())
+	}
+	for _, l := range source.lines {
+		g.lines = append(g.lines, l.dup())
+	}
+}
+
+func NewUnitFile() *UnitFile {
+	f := &UnitFile{
+		groups:      make([]*unitGroup, 0),
+		groupByName: make(map[string]*unitGroup),
+	}
+
+	return f
+}
+
+func ParseUnitFile(pathName string) (*UnitFile, error) {
```

Comment:
I wonder if we can simplify the parser.  Couldn't we require the `.container` files to start with the `[container]` table?  This way, Quadlet only had to parse the `[container]` table and leave the rest untouched (i.e., by parsing until EOF or the next table).

It seems to be possible to specify the same table multiple times, so I _imagine_:
```
[container]
...
[unit from user]
``` 
can be turned into
```
[unit from user]
[unit from quadlet]
[service from quadlet]
```

I think this could simplify the code quite a bit.

Reply:
I mean, i guess I see how we can e.g. allow the user to say:
```
[service]
Restart=always
```
And then when can set the exec line of the service section in a separate copy of it. However, we still do things like read the install section. And anyway, if we can parse one section, we could parse any? Is it really easier to parse one?